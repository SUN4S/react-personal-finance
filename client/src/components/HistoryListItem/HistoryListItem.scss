.historyListElement {
  width: 100%;
  min-height: 38px;
  max-height: 66px;
  display: flex;
  align-items: center;
  margin: 0;
  text-align: center;
  font-size: 16px;
  font-weight: 400;
  line-height: 24px;
  border-bottom: 2px solid var(--page-bg);
  color: var(--text-primary);
  overflow: hidden;
  animation: show 600ms 100ms cubic-bezier(0.38, 0.97, 0.56, 0.76) forwards;
  // Prestate
  opacity: 0;
  // remove transform for just a fade-in
  transform: rotateX(-90deg);
  transform-origin: top center;

  @keyframes show {
    100% {
      opacity: 1;
      transform: none;
    }
  }

  &:last-of-type {
    border: none;
  }

  .historyListElementDate {
    flex: 1;
  }
  .historyListElementCategory {
    flex: 2;
  }
  .historyListElementAmount {
    flex: 2;
  }
  .historyListElementDescription {
    flex: 4;
  }
  .historyListElementTags {
    flex: 2;
    max-height: 64px;
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    align-content: flex-start;
    overflow: hidden;

    .tagElement {
      min-width: 0;
      font-size: 12px;
      padding: 0 8px;
      margin: 2px 0;
      border-radius: 90px;
      border: 2px solid var(--primary);
    }
  }
  .historyListElementControls {
    flex: 1;

    button {
      height: 32px;
      min-width: 32px;
      width: 50%;
      display: flex;
      align-items: center;
      justify-content: center;
      margin: 0 auto;
      background-color: transparent;
      color: var(--text-primary);

      svg {
        height: 18px;
        width: 18px;
      }
    }
  }

  @media (max-width: 1200px) {
    .historyListElementTags {
      .tagElement {
        font-size: 10px;
      }
    }
  }

  @media (max-width: 768px) {
    .historyListElementDescription {
      display: none;
    }
    .historyListElementTags {
      display: none;
    }
  }

  @media (max-width: 360px) {
    font-size: 12px;
  }
}
