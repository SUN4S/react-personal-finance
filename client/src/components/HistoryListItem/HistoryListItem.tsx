import "./HistoryListItem.scss";

import { DateTime } from "luxon";
import { ExpenseState } from "../../models/expenses";
import { IconImage } from "../../resources/icons/IconImage/IconImage";

export const HistoryListItem = (props: {
  historyData: ExpenseState;
  clickFunction: Function;
}) => {
  return (
    <div className="historyListElement" data-testid="historyItem">
      <div className="historyListElementDate">
        {`${DateTime.fromISO(props.historyData.date).toFormat("MM-dd")}`}
      </div>
      <div className="historyListElementCategory">
        {props.historyData.category}
      </div>
      <div className="historyListElementAmount">{props.historyData.amount}</div>
      <div className="historyListElementDescription">
        {props.historyData.description
          ? props.historyData.description
          : "No Description"}
      </div>
      <div className="historyListElementTags">
        {props.historyData.tags &&
          props.historyData.tags.map((tag: string, index: number) => {
            return (
              <div className="tagElement" key={index}>
                {tag}
              </div>
            );
          })}
      </div>
      <div className="historyListElementControls">
        {props.historyData.receipt ? (
          <button
            aria-label="Preview Image"
            data-testid="previewImageButton"
            onClick={() => props.clickFunction()}
          >
            <IconImage />
          </button>
        ) : (
          "-"
        )}
      </div>
    </div>
  );
};
