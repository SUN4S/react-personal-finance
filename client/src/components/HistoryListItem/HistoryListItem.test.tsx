import { cleanup, fireEvent, render } from "@testing-library/react";

import { HistoryListItem } from "./HistoryListItem";
import renderer from "react-test-renderer";

describe("Expense List Item", () => {
  afterEach(cleanup);

  const mockFunction = jest.fn();
  const mockData = {
    _id: "62b1a3df9525535b9a48136b",
    category: "Culture",
    amount: 100,
    date: "2022-06-21T10:56:14.942+00:00",
    tags: ["Cinema", "Dog"],
    description: "Random description",
    receipt: "someImage.jpg",
  };

  it("Renders", () => {
    const { getByTestId } = render(
      <HistoryListItem historyData={mockData} clickFunction={mockFunction} />
    );
    expect(getByTestId("historyItem")).toBeTruthy();
  });

  it("Clicked Edit", () => {
    const { getByTestId } = render(
      <HistoryListItem historyData={mockData} clickFunction={mockFunction} />
    );
    fireEvent.click(getByTestId("previewImageButton"));
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });

  it("Renders Date", () => {
    const { getByTestId } = render(
      <HistoryListItem historyData={mockData} clickFunction={mockFunction} />
    );
    expect(
      getByTestId("historyItem").querySelector(".historyListElementDate")
    ).toHaveTextContent("06-21");
  });

  it("Renders Category", () => {
    const { getByTestId } = render(
      <HistoryListItem historyData={mockData} clickFunction={mockFunction} />
    );
    expect(
      getByTestId("historyItem").querySelector(".historyListElementCategory")
    ).toHaveTextContent("Culture");
  });

  it("Renders Amount", () => {
    const { getByTestId } = render(
      <HistoryListItem historyData={mockData} clickFunction={mockFunction} />
    );
    expect(
      getByTestId("historyItem").querySelector(".historyListElementAmount")
    ).toHaveTextContent("100");
  });

  it("Renders Description", () => {
    const { getByTestId } = render(
      <HistoryListItem historyData={mockData} clickFunction={mockFunction} />
    );
    expect(
      getByTestId("historyItem").querySelector(".historyListElementDescription")
    ).toHaveTextContent("Random description");
  });

  it("Renders Tags", () => {
    const { getByTestId } = render(
      <HistoryListItem historyData={mockData} clickFunction={mockFunction} />
    );
    expect(
      getByTestId("historyItem").querySelectorAll(".tagElement")[0]
    ).toHaveTextContent("Cinema");
  });

  it("Matches Snapshot", () => {
    const tree = renderer.create(
      <HistoryListItem historyData={mockData} clickFunction={mockFunction} />
    );
    expect(tree).toMatchSnapshot();
  });
});
