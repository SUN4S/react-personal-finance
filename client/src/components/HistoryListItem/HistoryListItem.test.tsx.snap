// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense List Item Matches Snapshot 1`] = `
<div
  className="historyListElement"
  data-testid="historyItem"
>
  <div
    className="historyListElementDate"
  >
    06-21
  </div>
  <div
    className="historyListElementCategory"
  >
    Culture
  </div>
  <div
    className="historyListElementAmount"
  >
    100
  </div>
  <div
    className="historyListElementDescription"
  >
    Random description
  </div>
  <div
    className="historyListElementTags"
  >
    <div
      className="tagElement"
    >
      Cinema
    </div>
    <div
      className="tagElement"
    >
      Dog
    </div>
  </div>
  <div
    className="historyListElementControls"
  >
    <button
      aria-label="Preview Image"
      data-testid="previewImageButton"
      onClick={[Function]}
    >
      <svg
        id="imageIcon"
        viewBox="0 0 52 52"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          data-name="Layer 2"
          id="Layer_2"
        >
          <g
            data-name="Layer 1"
            id="Layer_1-2"
          >
            <rect
              className="cls-1"
              height="50"
              rx="6"
              width="50"
              x="1"
              y="1"
            />
            <path
              className="cls-2"
              d="M46,44.07H6a1.46,1.46,0,0,1-1.1-2.18l9.54-18.82a1.21,1.21,0,0,1,2.22,0L22.1,33.81l9.35-18.42a1.22,1.22,0,0,1,2.22,0l13.44,26.5A1.46,1.46,0,0,1,46,44.07Z"
            />
            <circle
              className="cls-2"
              cx="10.24"
              cy="10.18"
              r="3.45"
            />
          </g>
        </g>
      </svg>
    </button>
  </div>
</div>
`;
